### ISSUE TITLE
Get rid of fromInternal/toInternal functions in favor of generics

### ISSUE THREAD

**Issue by anmolitor** [anmolitor/protoc-gen-elm#329]
I have recently thought about our current solution to handle mutually recursive protobuf modules/namespaces.

The examples below all reference the following protobuf file

```
// file: test.proto
package test;

enum Outer { A = 0 }

message Scope {
  enum Inner { B = 0 }
  message InnerMsg {
    Outer outer = 1;
    Inner inner = 2;
  }

  InnerMsg msg = 1;
}
```

The original protoc-gen-elm would have generated one big module:

```
type Test__Outer = Test__Outer__A
type Test__Scope__Inner = Test__Scope__Inner__B
type alias Test__Scope__InnerMsg = { outer : Test__Outer , inner : Test__Scope__Inner }
type alias Test__Scope = { msg : Test__Scope__InnerMsg }
```

We moved from this one big module to one module per protobuf namespace due to ugly names mostly - 
pattern matching on Test__Outer__Inner is very verbose for example. So the current implementation generates a module for the top level namespace and a seperate module for each inner namespace, for example the inside of `message Scope`.

```
// file: Test.elm
import Test.Scope

type Outer = A

type alias Scope = {
  inner : Test.Scope.InnerMsg
}
```

```
// file: Test/Scope.elm
import Test

type Inner = B

type alias InnerMsg = {
  outer : Test.Outer,
  inner : Inner
}
```

Except this does not work because the modules are mutually recursive and the Elm compiler does not allow that.
So we do something similar, but we use an _Internals module as the big file just like previously since a single file does not have issues with mutual recursion.
This approach comes with the disadvantage of generating the same union type multiple times. In contrast to type aliases for structs, we cannot rename the union type constructors. So we need to generate mapping functions between the union types which is not really user friendly at all, since the user just gets a type error with no explanation how to fix it.

To mitigate this I had the following idea:
What if we generated:

```
// file: Internals.elm
import Test.Unions
import Test.Scope.Unions

type alias Test__Outer = Test.Unions.Outer
type alias Test__Scope__Inner = Test.Scope.Unions.Inner
type alias Test__Scope__InnerMsg = { outer : Test__Outer , inner : Test__Scope__Inner }
type alias Test__Scope = { msg : Test__Scope__InnerMsg }
```

```
// file: Test/Scope.elm
import Internals
import Test.Scope.Unions

type alias Inner = Test.Scope.Unions.Inner

type alias InnerMsg = Internals.Test__Scope__InnerMsg 
```

```
// file: Test.elm
import Internals
import Test.Unions

type alias Outer = Test.Unions.Outer

type alias Scope = Internals.Test__Scope
```

```
// file: Test/Scope/Unions.elm
type Inner = B
```

```
// file: Test/Unions.elm
type Outer = A
```

So far so good, now when you want to pattern match on one of the generated union types, you need to import the .Unions file of the corresponding File you are importing as well, which is likely fine. We might even generate one file per union type so you can do

```
import Test.Scope.Inner as Inner
import Test.Scope exposing (Inner)

x : Inner
x = Inner.B
```

An issue arises when `oneof` declarations enter the mix.
Consider the following proto file

```
package order;
message Succeeded {
  string id = 1;
  string description = 2;
}
message Failed {
  string error = 1;
}
message OrderStatus {
  oneof status {
    Succeeded succeeded = 1;
    Failed failed = 2;
  } 
}
```

The oneof declaration generates a file `Order.OrderStatus.Status` which declares this union type:
```
type Status =
   Succeeded Internals.Order__Succeeded
  | Failed Internals.Order__Failed
```

and Internals, as usual has another declaration of the same union type:

```
// Internals.elm
type Order__OrderStatus__Status =
   Order__OrderStatus__Status __Succeeded Order__Succeeded
  | Order__OrderStatus__Status__Failed Order__Failed
```

We would like to just write

```
// Internals.elm
type alias Order__OrderStatus__Status = Order.OrderStatus.Status.Status
```

but we cannot since the module `Order.OrderStatus.Status` imports from `Internals`.

I propose the following solution:

```
// Order.OrderStatus.Status.elm
type Status a b =
   Succeeded a
  | Failed b
```

```
// Internals.elm
type alias Order__OrderStatus__Status = Order.OrderStatus.Status.Status Order__Succeeded Order__Failed
```

By introducing a generic parameter, we have effectively sidestepped the problem.
We can also generate

```
// Order.OrderStatus.elm
type alias Status = Order.OrderStatus.Status.Status Succeeded Failed
```

which should make the usage easier.
I have checked, and Elm does not seem to limit the amount of type parameters a type can have so this should scale without issues as well.

**Comment by anmolitor** [comment]
One issue I ran into when trying to implement this were recursive oneof types such as this one:

```
message LayerOne {
  repeated LayerTwo layerTwo = 1;
}

message LayerTwo {
  oneof msg {
    LayerOne layerOne = 1;
  }
}
```

Previously, there was no recursion issue here since we generated

```
type Msg = LayerOne LayerOne
```

which stops the recursion since it is not a type alias.

Now we want to generate

```
// Package.Msg.elm
type Msg a = LayerOne a
// Package.elm
type alias Msg = Msg LayerOne
// this expands into type alias Msg = Msg { msg: List LayerTwo }
// expands into type alias Msg = Msg { msg: List (Msg LayerOne ) }
// etc.
```

So similarily to directly recursive declarations like this one

```
message Rec {
  oneof msg {
    Rec rec = 1;
    string other = 2;
  }
}
```

we need to generate wrapper types and `wrap`/`unwrap` functions.

**Comment by anmolitor** [comment]
This feature is released in a 4.0.0-beta.0 version :)

**Comment by anmolitor** [comment]
Alright, I released 4.0 which generates this kind of code :)
